### PR TITLE
[DO NOT MERGE] Unmerged changes in v0.10.0-RC branch

### PR DESCRIPTION
**Pull request by rhornung67**
# Summary

- This PR compares the unmerged changes in v0.10.0-RC branch  to develop.
- It includes version updates and updated host-configs and CI updates for new versions of TPLs .

**Please review this PR, but approve https://github.com/LLNL/axom/pull/1425 for merging.**

### CODE REVIEW

**Comment by agcapps** [comment]
With commit 3f1fa72, I successfully built TPLs using vcpkg on my laptop, built Axom (both Debug and Release), and ran the test suite with no failures.

Note that this is without MPI and without Fortran, but with OpenMP.

**Comment by agcapps** [comment]
woops, there was a test failure on Release: slic_macros.

**Comment by bmhan12** [comment]
> woops, there was a test failure on Release: slic_macros.

Hm, I did run into an error on Windows with removing files in Azure Pipelines, and tried to guard against it:
https://github.com/LLNL/axom/blob/842a62463b413493bc459e52274c9c7c5d16a6d9/src/axom/slic/tests/slic_macros.cpp#L470-L474

Is it possibly related to that, or something else?

**Comment by agcapps** [comment]
> > woops, there was a test failure on Release: slic_macros.
> 
> Hm, I did run into an error on Windows with removing files in Azure Pipelines, and tried to guard against it:
> 
> https://github.com/LLNL/axom/blob/842a62463b413493bc459e52274c9c7c5d16a6d9/src/axom/slic/tests/slic_macros.cpp#L470-L474
> 
> Is it possibly related to that, or something else?

It is certainly related.  The errors I see are from that test: the files already exist (from a previous run) because they were not cleaned up.  If I delete file_no_fmt.txt and file_with_fmt.txt by hand (i.e. do the job from lines 472 and 473 that you disabled on Windows), the test succeeds and leaves those files behind.

**Comment by bmhan12** [comment]
> It is certainly related. The errors I see are from that test: the files already exist (from a previous run) because they were not cleaned up. If I delete file_no_fmt.txt and file_with_fmt.txt by hand (i.e. do the job from lines 472 and 473 that you disabled on Windows), the test succeeds and leaves those files behind.

Darn. Removing the check worked for Azure because every job run is a fresh build. Just realizing makes sense it would fail running the unit tests again. I'll try to debug this in the Windows Azure build, should hopefully fix things on your desktop as well.

**Comment by agcapps** [comment]
> > It is certainly related. The errors I see are from that test: the files already exist (from a previous run) because they were not cleaned up. If I delete file_no_fmt.txt and file_with_fmt.txt by hand (i.e. do the job from lines 472 and 473 that you disabled on Windows), the test succeeds and leaves those files behind.
> 
> Darn. Removing the check worked for Azure because every job run is a fresh build. Just realizing makes sense it would fail running the unit tests again. I'll try to debug this in the Windows Azure build, should hopefully fix things on your desktop as well.

What is happening is that the program still has handles open to the file: they haven't been closed.  On Windows, that causes the _unlink to fail.  On Linux, it doesn't care: it just deletes the file.  If you want to delete a file on Windows you have to close all open handles.

**Comment by bmhan12** [comment]
> What is happening is that the program still has handles open to the file: they haven't been closed. I confirmed this by running Process Explorer.
🤦  I didn't close the files prior to attempting to delete. Trying that now.

**Comment by rhornung67** [comment]
@bmhan12 if you have changes to make related to closing files, feel free to do that on this RC branch or on a separate PR. Whichever you prefer.

**Comment by agcapps** [comment]
So in order for `axom::utilities::filesystem::Unlink()` https://github.com/LLNL/axom/blob/develop/src/axom/core/utilities/FileUtilities.cpp#L126 to succeed, you will need to first call `std::ofstream::close()` wherever applicable.

I see that GenericOutputStream is built to deal with ostream, which is an abstraction level above ofstream.  The ofstream gets closed in its destructor (https://en.cppreference.com/w/cpp/io/basic_ofstream/close, see "Notes") and the GenericOutputStream deletes its ostream in ~GenericOutputStream, so before deleting the files, we could tell slic to delete all output streams.  @white238 , what do you think?

**Review by agcapps — APPROVED** [review]

File: host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake [hunk 1]
```diff
@@ -83,13 +83,13 @@ set(CUDA_TOOLKIT_ROOT_DIR "/usr/tce/packages/cuda/cuda-11.2.0" CACHE PATH "")
...
+set(CMAKE_CUDA_FLAGS "" CACHE STRING "")
+
 set(ENABLE_CUDA ON CACHE BOOL "")
 
 set(CMAKE_CUDA_SEPARABLE_COMPILATION ON CACHE BOOL "")
 
-set(CMAKE_CUDA_ARCHITECTURES "70" CACHE STRING "")
-
-set(CMAKE_CUDA_FLAGS "-restrict --expt-extended-lambda " CACHE STRING "")
+set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -restrict --expt-extended-lambda " CACHE STRING "" FORCE)
```
Comment (kennyweiss) [inline host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake:92]: Do we know why there are now two lines for `CMAKE_CUDA_FLAGS`?

I'm pretty sure the second `CMAKE_CUDA_FLAGS` w/ `FORCE` will overwrite the first one, so it's not a problem, but it does seem a bit weird to have both.

File: RELEASE-NOTES.md [hunk 2]
```diff
@@ -19,6 +19,21 @@ The Axom project release numbers follow [Semantic Versioning](http://semver.org/
...
 ## [Unreleased] - Release date yyyy-mm-dd
 
+### Added
+
+### Changed
+
+### Removed
+
+
+## [Version 0.10.0] - Release date 2024-09-xx
```
Comment (kennyweiss) [inline RELEASE-NOTES.md:29]: Please update the links for `axom@0.10` and `axom@0.9` at the bottom of this file

File: RELEASE-NOTES.md [hunk 3]
```diff
@@ -19,6 +19,21 @@ The Axom project release numbers follow [Semantic Versioning](http://semver.org/
...
+
+### Changed
+- Updates to [Conduit version 0.9.2][https://github.com/LLNL/conduit/releases/tag/v0.9.2]
+- Updates to [RAJA version 2024.07.0][https://github.com/LLNL/RAJA/releases/tag/v2024.07.0]
+- Updates to [camp version 2024.07.0][https://github.com/LLNL/camp/releases/tag/v2024.07.0]
+- Updates to [Umpire version 2024.07.0][https://github.com/LLNL/Umpire/releases/tag/v2024.07.0]
```
Comment (kennyweiss) [inline RELEASE-NOTES.md:35]: Please merge these four lines w/ the existing `Changed` section for axom@0.10 a few lines down (below `Added`)

**Review by kennyweiss — COMMENTED** [review]
Looks good, but I'd like to take a pass through the RELEASE_NOTES before the release.
Also, should we add a line to the `RELEASE_NOTES` about dropping support for XL?

(Note: I'm intentionally not approving this to ensure the branch doesn't get merged into develop)

File: host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake (hunk 1 above)
Comment (rhornung67) [reply host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake:92]: I have no idea why there are two lines for `CMAKE_CUDA_FLAGS`. I just ran the scripts to generate the new host-configs.

**Comment by kennyweiss** [comment]
> So in order for `axom::utilities::filesystem::Unlink()` https://github.com/LLNL/axom/blob/develop/src/axom/core/utilities/FileUtilities.cpp#L126 to succeed, you will need to first call `std::ofstream::close()` wherever applicable.
> 
> I see that GenericOutputStream is built to deal with ostream, which is an abstraction level above ofstream. The ofstream gets closed in its destructor (https://en.cppreference.com/w/cpp/io/basic_ofstream/close, see "Notes") and the GenericOutputStream deletes its ostream in ~GenericOutputStream, so before deleting the files, we could tell slic to delete all output streams. @white238 , what do you think?

Really nice find @agcapps!
Can we please add a note about this in the doxygen for `removeFile`?
https://github.com/LLNL/axom/blob/24751dd2a06152f0bb6d057377b0674c837b0390/src/axom/core/utilities/FileUtilities.hpp#L86-L92

@bmhan12 -- perhaps this should be added to #1424 ?

File: host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake (hunk 1 above)
Comment (kennyweiss) [reply host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake:92]: It's likely related to the class hierarchy for packages in Spack.
Thoughts @white238 ?

Comment (white238) [reply host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake:92]: There were some changes in spack. This was reviewed in a previous PR.

File: RELEASE-NOTES.md (hunk 3 above)
Comment (rhornung67) [reply RELEASE-NOTES.md:35]: I don't understand. This follows what is done for other releases that have updated TPL versions.

Comment (kennyweiss) [reply RELEASE-NOTES.md:35]: There are two `Changed` sections for this release -- https://github.com/LLNL/axom/blob/b1f5bbc7c617faaf28643f0f88184ccc7d52eae2/RELEASE-NOTES.md

I can take care of the fix

File: RELEASE-NOTES.md (hunk 2 above)
Comment (rhornung67) [reply RELEASE-NOTES.md]: Done.

File: host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake (hunk 1 above)
Comment (rhornung67) [reply host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang@10.0.1.2_cuda.cmake:92]: A note about no longer supporting XL has been added to the release notes.

File: RELEASE-NOTES.md (hunk 3 above)
Comment (rhornung67) [reply RELEASE-NOTES.md:35]: I see what you mean now. I thought the TPL version updates should be up front so they are more obvious.  But, whatever you prefer is fine with me.

File: RELEASE-NOTES.md [hunk 4]
```diff
@@ -17,10 +17,10 @@ The format of this file is based on [Keep a Changelog](http://keepachangelog.com
 
 The Axom project release numbers follow [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
-## [Unreleased] - Release date yyyy-mm-dd
```
Comment (kennyweiss) [inline RELEASE-NOTES.md:20]: @rhornung67 -- I removed the empty `Unreleased` section for the release. We'll add it back when we merge main to develop

**Review by white238 — APPROVED** [review]

**Pull request closed without merging**